### PR TITLE
[FLINK-26694][table] Support lookup join via a multi-level inheritance of TableFunction

### PR DESCRIPTION
**Pull request by YesOrNo828**
## What is the purpose of the change

Support Flink SQL lookup joins via a multi-level inheritance of TableFunction.

For example:
```
    private static class LookupTableLevel0Function<T> extends TableFunction<T> {}

    public static class LookupTableWithoutHintLevel1Function
            extends LookupTableLevel0Function<RowData> {
        public void eval(@DataTypeHint("STRING") StringData s) {...}
    }
```
Lookup Joins could be runable when LookupTableSource return `TableFunctionProvider.of(LookupTableWithoutHintLevel1Function)`

follow up: #19124, and add some tests to cover the change

## Brief change log

  - *Using the `Class.isAssignableFrom` to detemine the baseClass(TableFunction) is superclass of clazz(user custom TableFunction)*
  - *Add some tests to cover the change*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unit test testLookupTableFunctionWithoutHintLevel1 for multi-level inheritance of TableFunction*
  - *Added unit test testLookupTableFunctionWithoutHintLevel0 for one level inheritance of TableFunction*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no) no 
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no) no
  - The serializers: (yes / no / don't know) no
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know) no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know) no
  - The S3 file system connector: (yes / no / don't know) no

## Documentation

  - Does this pull request introduce a new feature? (yes / no) no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by YesOrNo828** [comment]
@flinkbot run azure

**Comment by YesOrNo828** [comment]
@flinkbot run azure

**Comment by YesOrNo828** [comment]
@flinkbot run azure

**Comment by YesOrNo828** [comment]
@flinkbot run azure

**Comment by YesOrNo828** [comment]
@flinkbot run azure

**Comment by YesOrNo828** [comment]
@leonardBang  @lincoln-lil  Could you review this patch when you have time? Thanks.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java [hunk 1]
```diff
@@ -1698,6 +1717,32 @@ public void eval(@DataTypeHint("STRING") StringData s) {
...
     }
 
+    /** This is an empty synchronous table function. */
+    private static class LookupTableLevel0Function<T> extends TableFunction<T> {}
+
+    /**
+     * Synchronous table function that uses {@link RowData} type inference for {@link
+     * LookupTableSource} and inherits {@link TableFunction} at multiple levels.
+     */
+    public static class LookupTableWithoutHintLevel1Function
```
Comment (xuyangzhong) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java:1727]: Just curious, should we add a test like LookupTableWithHintLevel1Function for considerations about the test orthogonality?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java [hunk 2]
```diff
@@ -1698,6 +1717,32 @@ public void eval(@DataTypeHint("STRING") StringData s) {
         }
     }
 
+    /** This is an empty synchronous table function. */
+    private static class LookupTableLevel0Function<T> extends TableFunction<T> {}
+
+    /**
+     * Synchronous table function that uses {@link RowData} type inference for {@link
```
Comment (xuyangzhong) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java:1724]: It seems not to use the RowData type inference, right?

**Review by xuyangzhong — COMMENTED** [review]
Hi, thanks for your contribution. This pr LGTM overall. I just left some tiny comments here.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java (hunk 1 above)
Comment (YesOrNo828) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java:1740]: I have added the `testLookupTableFunctionWithHintLevel1` test.

**Comment by YesOrNo828** [comment]
@xuyangzhong  Thanks for your review. I have rebased and added the `testLookupTableFunctionWithHintLevel1` test. PTAL.

**Review by xuyangzhong — APPROVED** [review]
LGTM, cc @leonardBang for the final check.

**Comment by xuyangzhong** [comment]
BTW, the CI seems failed. Can you take a look @YesOrNo828 ?

**Comment by YesOrNo828** [comment]
@flinkbot run azure

**Comment by leonardBang** [comment]
@YesOrNo828 I'll take a look soon.

**Review by leonardBang — APPROVED** [review]
Thanks @YesOrNo828 for the contribution and @xuyangzhong for the review, LGTM

**Pull request merged**